src/compiler/scala/tools/nsc/javac/JavaParsers.scala

```diff
@@ -326,20 +336,55 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
     }
 
     def annotations(): List[Tree] = {
-      //var annots = new ListBuffer[Tree]
+      var annots = new ListBuffer[Tree]
       while (in.token == AT) {
         in.nextToken()
-        annotation()
+        annots += annotation()
       }
-      List() // don't pass on annotations for now
+      annots.toList
     }
 
     /** Annotation ::= TypeName [`(` AnnotationArgument {`,` AnnotationArgument} `)`]
      */
-    def annotation() {
-      qualId()
-      if (in.token == LPAREN) { skipAhead(); accept(RPAREN) }
-      else if (in.token == LBRACE) { skipAhead(); accept(RBRACE) }
+    def annotation(): Tree = {
+      def annArg(): Tree = {
+        def annVal(): Tree = {
```

Comment:
This method is too restrictive. The [Java syntax](http://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.7) allows `ConditionalExpression` here, which can be many things. For example, the following compiles with javac, but gives a syntax error with this PR:

```
➜  sandbox git:(be39803acc) ✗ cat A.java
public @interface A {
  int value();
}
class T {
  public static final int x = 1;
  @A(1 + x) void t2() { }
}
➜  sandbox git:(be39803acc) ✗ qsc A.java
A.java:6: error: `)' expected but `+' found.
  @A(1 + x) void t2() { }
       ^
A.java:7: error: illegal start of type
}
^
two errors found
```

If we make the field `x` non-final, it's no longer a compile-time constant, and the Java compiler shows an error "element value must be a constant expression". However, this is not a syntax error.

Supporting all valid Java annotations would require a huge amount of work: we'd have to start parsing Java expressions and constant fold them. In the Scala compiler this is currently type-driven (ConstantType), but in the case of java we already have a type for compile-time constants (`x` is `int` above). See also an older comment here: https://github.com/scala/bug/issues/5333#issuecomment-292401305

However, this should not stop us from supporting basic Java annotations. But we have to make sure that the Scala compiler doesn't reject Java code that the Java compiler accepts.